pypeit/coadd3d.py

```diff
@@ -0,0 +1,1772 @@
+                cd_wv = self.cubepar['wave_delta'] if self.cubepar['wave_delta'] is not None else 1.0E10 * frame_wcs.wcs.cd[2, 2]
+                output_wcs = self.spec.get_wcs(spec2DObj.head0, slits, detector.platescale, crval_wv, cd_wv)
+                # Set the wavelength range of the white light image.
+                wl_wvrng = None
+                if self.cubepar['save_whitelight']:
+                    wl_wvrng = datacube.get_whitelight_range(np.max(self.mnmx_wv[ff, :, 0]),
+                                                    np.min(self.mnmx_wv[ff, :, 1]),
+                                                    self.cubepar['whitelight_range'])
+                # Make the datacube
+                if self.method in ['subpixel', 'ngp']:
+                    # Generate the datacube
+                    generate_cube_subpixel(outfile, output_wcs, ra_sort[resrt], dec_sort[resrt], wave_sort[resrt],
+                                           flux_sort[resrt], ivar_sort[resrt], np.ones(numpix),
+                                           this_spatpos, this_specpos, this_spatid,
+                                           spec2DObj.tilts, slits, alignSplines, bins,
+                                           all_idx=None, overwrite=self.overwrite,
+                                           blaze_wave=self.blaze_wave, blaze_spec=self.blaze_spec,
+                                           fluxcal=self.fluxcal, specname=self.specname, whitelight_range=wl_wvrng,
+                                           spec_subpixel=self.spec_subpixel, spat_subpixel=self.spat_subpixel)
+                continue
+
+            # Store the information if we are combining multiple frames
+            self.all_ra = np.append(self.all_ra, ra_sort[resrt])
+            self.all_dec = np.append(self.all_dec, dec_sort[resrt])
+            self.all_wave = np.append(self.all_wave, wave_sort[resrt])
+            self.all_sci = np.append(self.all_sci, flux_sort[resrt])
+            self.all_ivar = np.append(self.all_ivar, ivar_sort[resrt].copy())
+            self.all_idx = np.append(self.all_idx, ff * np.ones(numpix))
+            self.all_wghts = np.append(self.all_wghts, self.weights[ff] * np.ones(numpix) / self.weights[0])
+            self.all_spatpos = np.append(self.all_spatpos, this_spatpos)
+            self.all_specpos = np.append(self.all_specpos, this_specpos)
+            self.all_spatid = np.append(self.all_spatid, this_spatid)
+            self.all_tilts.append(spec2DObj.tilts)
+            self.all_slits.append(slits)
+            self.all_align.append(alignSplines)
+
+    def run_align(self):
+        """
+        This routine aligns multiple cubes by using manual input offsets or by cross-correlating white light images.
+        """
+        # Grab cos(dec) for convenience
+        cosdec = np.cos(np.mean(self.all_dec) * np.pi / 180.0)
+
+        # Register spatial offsets between all frames
+        if self.opts['ra_offset'] is not None:
```

Comment:
Likewise, all of what I'm seeing here is core functionality. Classes should wrap core methods, and reduce the amount of replicated code exploiting inheritance, and other niceties of object oriented programming. What I see here is a gigantic hugely complex class, without the management of variables (local vs global) or control flow afforded by core functions.